pkg/liqonet/iptables/iptables.go

```diff
@@ -0,0 +1,668 @@
+		}
+		klog.Infof("Deleted chain %s (table %s)", chain, getTableFromChain(chain))
+	}
+	return nil
+}
+
+// EnsureChainRulesPerCluster reads TunnelEndpoint resource and
+// makes sure that chain rules for the given cluster exist.
+func (h IPTHandler) EnsureChainRulesPerCluster(tep *netv1alpha1.TunnelEndpoint) error {
+	chainRules, err := getChainRulesPerCluster(tep)
+	if err != nil {
+		return err
+	}
+	for chain, newRules := range chainRules {
+		// Invocking updateRulesPerChain would be an error
+		// since it would replace all the rules in a Liqo chain
+		// with rules related to this remote cluster.
+		// Therefore, first get existing rules related to remote cluster
+		// and then call updateSpecificRulesPerChain.
+		rules, err := h.getExistingChainRules(tep.Spec.ClusterID, chain)
+		if err != nil {
+			return fmt.Errorf("cannot get existing chain rules per cluster %s: %w", tep.Spec.ClusterID, err)
+		}
+		if err := h.updateSpecificRulesPerChain(chain, rules, newRules); err != nil {
+			return fmt.Errorf("cannot update rule for chain %s (table %s): %w", chain, getTableFromChain(chain), err)
+		}
+	}
+	return nil
+}
+
+func (h IPTHandler) getExistingChainRules(clusterID, chain string) ([]string, error) {
+	existingChainRules := make([]string, 0)
+	// Get rules in chain
+	rules, err := h.ListRulesInChain(chain)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list rules in chain %s (table %s): %w",
+			chain, getTableFromChain(chain), err)
+	}
+
+	// Get cluster chains related to received chain
+	clusterChains := getChainsPerCluster(clusterID)
+
+	// Check if an existing rule refers to a cluster chain
+	for _, rule := range rules {
+		for _, clusterChain := range clusterChains {
+			if !strings.Contains(rule, clusterChain) {
+				continue
+			}
+			// Rule refers to cluster chain
+			existingChainRules = append(existingChainRules, rule)
+			break
+		}
+	}
+	return existingChainRules, nil
+}
+
+func getChainsPerCluster(clusterID string) map[string]string {
+	chains := make(map[string]string)
+	chains[consts.ForwardChain] = getClusterForwardChain(clusterID)
+	chains[consts.InputChain] = getClusterInputChain(clusterID)
+	chains[consts.PostroutingChain] = getClusterPostRoutingChain(clusterID)
+	chains[consts.PreroutingChain] = getClusterPreRoutingChain(clusterID)
+	return chains
+}
+
+// EnsureChainsPerCluster is used to be sure input, output, postrouting and prerouting chain for a given
+// cluster are present in the NAT table and Filter table.
+func (h IPTHandler) EnsureChainsPerCluster(clusterID string) error {
+	if clusterID == "" {
+		return &errors.WrongParameter{
+			Parameter: consts.ClusterIDLabelName,
+			Reason:    errors.StringNotEmpty,
+		}
+	}
+	// Get chains per cluster
+	chains := getChainsPerCluster(clusterID)
+	for _, chain := range chains {
+		if err := h.createIptablesChainIfNotExists(getTableFromChain(chain), chain); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getTableFromChain(chain string) string {
```

Comment:
I would add a comment on this method since it is frequently called but it is not straightforward to get what should return.